[Windows] Replace Header with PlaceholderText for `Title` property of `Picker`

mhrastegari opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

<!--
!!!!!!! MAIN IS THE ONLY ACTIVE BRANCH. MAKE SURE THIS PR IS TARGETING MAIN. !!!!!!! 
-->

### Description of Change

This change updates the `Picker` for `Windows` so that its `Title` property matches the behavior on iOS and Android.
Instead of displaying the title as a header, it now uses it as the placeholder to ensure consistent cross-platform appearance.

### Platform
- [x] Windows

### Issues Fixed

Fixes #32723
Fixes #6845

### Screenshots

| Before | After |
| ------ | ----- |
| <img src="https://github.com/user-attachments/assets/bdae96ce-4a46-4188-b52b-3d2b0e12fc1c" width="500" /> | <img src="https://github.com/user-attachments/assets/93a3416a-5355-4dad-85a6-1e58b90f20fa" width="500" /> |


---

This is a breaking change and therefore targets the .NET 11 branch as requested.

mhrastegari committed:
replace Header with PlaceholderText for Title property of Picker on Windows (#32723)

MartyIX commented:
@mhrastegari Perhaps it's worth adding "Fixes #6845" to the OP.

kubaflo reviewed (changes requested):
Could you please add a UI Test?

mhrastegari committed:
add UI Tests

mhrastegari commented:
Sorry for the late response @kubaflo! Added the UI Tests.

kubaflo reviewed (changes requested):
Could you please review the ai's summary?

mhrastegari committed:
remove dead ComboBox styles

mhrastegari committed:
add UpdateTitleColor to PickerExtensions

mhrastegari committed:
add more tests

mhrastegari commented:
Done @kubaflo!

mhrastegari committed:
update unshipped public API text

mhrastegari committed:
Merge branch 'main' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

kubaflo commented:
/review

mhrastegari committed:
add more tests for TitleColor

mhrastegari committed:
add CharacterSpacing test

mhrastegari committed:
resolve MauiBot review comments

kubaflo committed:
Merge branch 'net11.0' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

mhrastegari committed:
fix CharacterSpacing test

mhrastegari committed:
resolve review comments

kubaflo reviewed (changes requested):
Could you please add ui tests?

mhrastegari committed:
add more tests

kubaflo reviewed (changes requested):
We also need UI Tests - could you please add them?

mhrastegari committed:
add Appium UI tests

kubaflo committed:
Merge branch 'net11.0' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

mhrastegari committed:
fix conflicts

mhrastegari committed:
Merge branch 'net11.0' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

mhrastegari committed:
resolve review comments

kubaflo reviewed (changes requested):
Looks like there are a few tests regressions - could you please verify? Maybe you just need to update snapshots :)

kubaflo reviewed (changes requested):
Looks like there are test failures

mhrastegari committed:
resolve review comments

mhrastegari committed:
update screenshots

mhrastegari committed:
resolve review comments

kubaflo reviewed (changes requested):
Could you please check the latest ai's suggestions?

kubaflo reviewed (changes requested):
Looks like the test keeps on failing

mhrastegari committed:
resolve review comments

kubaflo reviewed (changes requested):
Could you please resolve conflicts and check the ai's suggestions?

mhrastegari committed:
Merge branch 'net11.0' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

Resolved conflicts:
- PickerExtensions.cs: kept both CharacterSpacing paths. Upstream #30612 applies
  spacing to the selected item's TextBlock; this branch applies it to the
  placeholder TextBlock that now renders Title. Both are needed.
- Styles/MauiComboBoxStyle.xaml: kept this branch's deletion. The file only held
  the ComboBoxHeader template, which is obsolete now that Title maps to the
  native ComboBox.PlaceholderText.
- PublicAPI/net-windows/PublicAPI.Unshipped.txt: kept both added API entries.

mhrastegari committed:
resolve review comments

- Apply CharacterSpacing to the placeholder and the selected item from a single
  deferred callback instead of two separate Loaded subscriptions.
- Refresh theme resources after TitleColor changes so the placeholder picks up
  the new brush in every ComboBox visual state.

mhrastegari committed:
update screenshots

Regenerate the Issue30464 Windows baseline. Its Picker Title now renders as
placeholder text inside the ComboBox instead of a header above it, so the
snapshot added by #30612 no longer matched.

kubaflo reviewed (changes requested):
There's a couple of ui test failures: https://github.com/dotnet/maui/pull/33007#issuecomment-5078348984

mhrastegari committed:
update screenshots

kubaflo committed:
Merge branch 'net11.0' into 32723-replace-header-with-placeholder-on-windows-picker-title-property

kubaflo commented on src/Core/src/Platform/Windows/PickerExtensions.cs:
```diff
@@ -10,11 +9,18 @@ public static class PickerExtensions
 		public static void UpdateTitle(this ComboBox nativeComboBox, IPicker picker)
 		{
-			nativeComboBox.Header = string.IsNullOrEmpty(picker.Title) ? null : picker;
+			nativeComboBox.PlaceholderText = picker.Title ?? string.Empty;
```
**[major] Regression Prevention and Test Coverage** — Windows snapshot baselines are only partially refreshed. Moving `Title` from `ComboBox.Header` to `PlaceholderText` removes the header label from *every* Windows Picker that has a `Title`, not just the tests whose names contain "Title". I extracted the baselines at this head and visually confirmed the following committed Windows PNGs still show the **old** header rendering and were **not** updated by this PR:

- `Picker_Validate_VerifyLabels.png` — shows `Picker Title` as a label above an empty ComboBox
- `Picker_SetSelectedIndex_VerifySelectedIndexAndItem.png` — header `Picker Title` above `Option 2 - Second option`
- `Picker_SetSelectedItem_VerifySelectedItemLabel.png`
- `Picker_SetFontSizeAndFontAttributesBold_VerifyFontSizeAndAttributes.png` — header `Picker Title` above the bold selected item
- `Picker_SetFontSizeAndFontFamilyDokdo_VerifyFontSizeAndFontFamily.png`
- `Picker_SetFlowDirectionRTL_VerifyFlowDirection.png` — right-aligned header `Picker Title`
- `PickerKeepsCharacterSpacingAfterSelectingItem.png` / `PickerKeepsCharacterSpacingAfterProgrammaticSelectedIndexChange.png` — header `S e l e c t   a n   I t e m` above the ComboBox
- `LightTheme_Picker_VerifyVisualState.png` / `DarkTheme_Picker_VerifyVisualState.png` — header `Select a monkey` above the ComboBox
- `Issue32984PickerShouldResize.png` — host page sets `Title = "Select an item"`, so the ComboBox height/dropdown anchor shifts

Root cause for the feature-matrix set: `FeatureMatrix/Picker/PickerControlPage.xaml` binds `Title="{Binding Title}"` and `PickerViewModel` defaults `_title = "Picker Title"`, so *all* `Picker_*` Windows baselines rendered a header before this change — not only the `*Title*`-named ones that were regenerated. Same for `FeatureMatrix/AppTheme/AppThemePage.xaml` / `DefaultThemePage.xaml` (`Title="Select a monkey"`).

Note this is a different (and complementary) set from the earlier bot comment on this line: the baselines that comment listed (`Picker_SetTitle_VerifyTitleLabel.png`, `PickerShouldRegainTitle.png`, `PickerTitleShouldBeRed.png`, etc.) *are* correctly updated at this head; the ones above are the remaining stale ones. Please regenerate them or the Windows UI test lane will stay red after merge.

kubaflo commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue32723.cs:
```diff
@@ -0,0 +1,35 @@
+	public void Issue32723PickerTitleUpdatesAtRuntime()
+	{
+		App.WaitForElement("issue32723Button");
+		App.Tap("issue32723Button");
```
**[moderate] Regression Prevention and Test Coverage** — This tap mutates persistent page state and nothing restores it between tests. The host page handler toggles both properties (`Issue32723.cs`: `picker.Title = picker.Title == "Select Option" ? "Updated Title" : "Select Option"` and the same for `TitleColor`), but this fixture does not override `ResetAfterEachTest`, which defaults to `false` (`src/TestUtils/src/UITest.NUnit/UITestBase.cs:12`), and `_IssuesUITest.TryToResetTestState()` re-navigates only from `FixtureSetup`/`OneTimeSetUp` — not per test.

Concrete failure: the two tests carry no `[Order]`, and NUnit does not guarantee method execution order within a fixture. If `Issue32723PickerTitleUpdatesAtRuntime` runs before `Issue32723PickerTitleDisplaysAsPlaceholder`, the latter captures the toggled state (`Updated Title` in blue) and fails against `Issue32723PickerTitleDisplaysAsPlaceholder.png`. Because the button is a *toggle* rather than an absolute setter, re-entering this test in the same app session also flips the picker back to the initial state and breaks `Issue32723PickerTitleUpdatesAtRuntime.png`.

Add `protected override bool ResetAfterEachTest => true;` (the established convention for state-mutating issue fixtures, e.g. `Issue28416.cs:15`, `Issue20834.cs:14`), and/or make the button assign an absolute `Title`/`TitleColor` instead of toggling.

kubaflo reviewed (commented):
Expert review — see inline comments.

kubaflo reviewed (changes requested):
https://github.com/dotnet/maui/pull/33007#issuecomment-5231779793

kubaflo commented:
Thanks @mhrastegari for all the work and responsiveness on this PR. After reviewing the product tradeoffs, I don’t think we should take this change, so I’m going to close it.

While cross-platform consistency is desirable, the current Windows behavior uses the native `ComboBox.Header` and is not a functional defect. Changing `Picker.Title` globally to `PlaceholderText` would be a breaking behavioral and layout change for every existing Windows app: the persistent title disappears after selection, and removing that persistent context also introduces accessibility risk.

The implementation has additionally exposed a disproportionate regression and maintenance surface around character spacing, theme states, handler lifecycle, and existing visual baselines. Given the relatively modest demand and the availability of an opt-in handler-mapper workaround, the benefit does not justify that compatibility cost.

This is a product decision rather than a reflection on the quality or effort of the contribution. If we revisit this area, I think it should be through an additive, opt-in placeholder API that preserves the existing Windows `Title` behavior rather than silently changing it. Thanks again for helping us explore the problem thoroughly.
